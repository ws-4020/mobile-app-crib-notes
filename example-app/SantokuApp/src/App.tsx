import {NavigationContainer} from '@react-navigation/native';
import type {ParamListBase} from '@react-navigation/routers';
import {activateKeepAwake} from 'expo-keep-awake';
import {RootStackNav} from 'navigation';
import React from 'react';

import {BundledMessagesLoader, loadMessages} from './framework';
import {firebaseConfig} from './framework/firebase';
import {log} from './framework/logging';

// React Navigation の useNavigation/Link/ref 等のデフォルトタイプを指定
declare global {
  namespace ReactNavigation {
    interface RootParamList extends ParamListBase {}
  }
}

export const App = () => {
  // 開発中は画面がスリープしないようにしておきます。
  if (__DEV__) {
    activateKeepAwake();
  }

  if (!firebaseConfig.isDummy) {
    // Firebase Crashlyticsの初期化
    require('@react-native-firebase/crashlytics');
  }
  // アプリ内で使用するメッセージのロード
  loadMessages(new BundledMessagesLoader()).catch(() => {
    // アプリにバンドルしているメッセージのロードは失敗しない想定
    log.error('Failed to load message.', 'BundledMessagesLoadingFailure');
  });

  return (
    <NavigationContainer>
      <RootStackNav />
    </NavigationContainer>
  );
};
