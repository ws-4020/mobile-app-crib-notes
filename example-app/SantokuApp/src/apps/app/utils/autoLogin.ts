import {isUnauthorizedError} from 'features/account/errors/UnauthorizedError';
import {autoLogin as accountAutoLogin} from 'features/account/utils/autoLogin';
import {clientLogout} from 'features/account/utils/clientLogout';

import {InitialDataError} from '../errors/initialDataError';

export const autoLogin = async () => {
  try {
    return await accountAutoLogin();
  } catch (e) {
    if (isUnauthorizedError(e)) {
      await clientLogout();
      throw e;
    }
    throw new InitialDataError(e);
  }
};
