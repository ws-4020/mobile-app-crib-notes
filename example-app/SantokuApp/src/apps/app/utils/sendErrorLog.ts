import axios from 'axios';
import {RuntimeError} from 'bases/core/error/RuntimeError';
import {log} from 'bases/logging';
import {ErrorResponse} from 'features/backend/apis/model';

export const sendErrorLog = (error: unknown) => {
  try {
    if (axios.isAxiosError(error)) {
      const data = error.response?.data as ErrorResponse | undefined;
      const errorCode = data?.code ?? 'AxiosError';
      log.error(error, errorCode);
    } else {
      if (error instanceof Error) {
        log.error(error, 'UnexpectedRequestError');
      } else {
        log.error(new RuntimeError(error), 'UnexpectedRequestError');
      }
    }
  } catch {}
};
