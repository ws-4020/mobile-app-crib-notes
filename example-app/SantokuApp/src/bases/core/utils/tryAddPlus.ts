/**
 * 入力された値が最大値より大きい場合、整形(最大値の末尾に+を付与)した値を返す。
 * 最大値より小さい場合、入力された値を返す。
 * @param value 整形したい値
 * @param maxValue 最大値
 * @returns 整形後の値
 */

export function tryAddPlus(value: number, maxValue: number): string {
  if (value <= maxValue) {
    return value.toString();
  }
  return maxValue.toString() + '+';
}
