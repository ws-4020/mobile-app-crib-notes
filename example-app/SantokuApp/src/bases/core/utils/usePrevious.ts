import {useEffect, useRef} from 'react';

/**
 * This React hook provides the previous value. (To be precise, it returns the previous reference.)
 *
 * We have previously suggested a custom Hook called usePrevious to hold the previous value.
 * However, we’ve found that most use cases fall into the two patterns described above.
 * If your use case is different, you can hold a value in a ref and manually update it when needed.
 * Avoid reading and updating refs during rendering because this makes your component’s behavior difficult to predict and understand.
 *
 * See [Storing information from previous renders](https://ja.react.dev/reference/react/useState#storing-information-from-previous-renders).
 */
export const usePrevious = <T>(current: T) => {
  const ref = useRef<T>();

  useEffect(() => {
    ref.current = current;
  });

  return ref.current;
};
