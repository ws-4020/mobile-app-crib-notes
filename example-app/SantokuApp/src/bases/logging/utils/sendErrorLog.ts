let sendErrorLog;

const setSendErrorLog = (delegate: (error: unknown) => void) => {
  sendErrorLog = delegate;
};

export {sendErrorLog, setSendErrorLog};
