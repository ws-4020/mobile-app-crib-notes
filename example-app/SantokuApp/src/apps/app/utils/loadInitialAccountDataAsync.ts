import {getAccountsMeTerms} from 'features/account/utils/auth/getAccountsMeTerms';
import {getAccountsMe} from 'features/backend/apis/account/account';
import {getTerms} from 'features/terms/utils/getTerms';

import {AccountData} from '../../../features/account/types/AccountData';
import {InitialDataError} from '../errors/initialDataError';

const loadInitialAccountDataAsync = async (): Promise<AccountData> => {
  try {
    const account = (await getAccountsMe()).data;
    const termsAgreementStatus = (await getAccountsMeTerms()).data;
    let termsOfService;
    if (!termsAgreementStatus.hasAgreed) {
      termsOfService = (await getTerms()).data;
    }
    return {
      account,
      terms: {termsAgreementStatus, termsOfService},
    };
  } catch (e) {
    throw new InitialDataError('Failed to load initial account data.', e);
  }
};

export {loadInitialAccountDataAsync};
