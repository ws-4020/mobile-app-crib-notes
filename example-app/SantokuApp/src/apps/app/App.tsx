import {setErrorSender} from 'bases/error/sendErrorLog';
import {firebaseConfig} from 'bases/firebase/configs/FirebaseConfig';
import {createLogger, setLogger} from 'bases/logging/utils';
import {FirebaseCrashlyticsTransport} from 'bases/logging/utils/FirebaseCrashlyticsTransport';
import React from 'react';
import {Platform, StyleSheet} from 'react-native';
import {GestureHandlerRootView} from 'react-native-gesture-handler';
import {SafeAreaProvider} from 'react-native-safe-area-context';

import {AppWithInitialization} from './AppWithInitialization';
import {WithAppTheme} from './providers/AppThemeContext';
import {WithOverlay} from './providers/WithOverlay';
import {sendErrorLog} from './utils/sendErrorLog';

type AppProperties = {
  [key: string]: any;
};

setLogger(
  __DEV__ || firebaseConfig.isDummy
    ? createLogger({level: 'trace'})
    : createLogger({level: 'error', transports: [new FirebaseCrashlyticsTransport()]}),
);

setErrorSender(sendErrorLog);

export const App = ({isHeadless}: AppProperties) => {
  // isHeadlessを取得するためにはAppDelegate.mに変更が必要
  // https://rnfirebase.io/messaging/usage#background-application-state
  if (Platform.OS === 'ios' && isHeadless) {
    // iOSでバックグラウンド時にリモート通知を受信するとアプリケーションがサイレントに起動される
    // この時にはReact Componentを返さないようにし、useEffectなどの処理がバックグラウンドで走ることを防止する
    // Androidの場合はApp自体が読み込まれないので追加対応は必要ない
    return null;
  }

  return (
    <GestureHandlerRootView style={StyleSheet.absoluteFill}>
      <SafeAreaProvider>
        <WithAppTheme>
          <WithOverlay>
            <AppWithInitialization />
          </WithOverlay>
        </WithAppTheme>
      </SafeAreaProvider>
    </GestureHandlerRootView>
  );
};
