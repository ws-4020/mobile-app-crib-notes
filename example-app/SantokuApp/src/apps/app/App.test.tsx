import * as Application from 'expo-application';
import * as SecureStore from 'expo-secure-store';
import '@testing-library/jest-native/extend-expect';
import {render, screen, waitFor} from '@testing-library/react-native';
import {BACKEND_AXIOS_INSTANCE_WITHOUT_REFRESH_SESSION} from 'features/backend/utils/customInstance';
import {initialData} from 'fixtures/msw/datas';
import {setupServer} from 'msw/node';
import React from 'react';

import {initialDb} from '../../fixtures/msw/db';
import {handlers} from '../../fixtures/msw/handlers';
import {App} from './App';

// jest.mock('react-native/Libraries/Utilities/DevSettings.js', () => {
//   return {addMenuItem: jest.fn};
// });
// jest.spyOn(BACKEND_AXIOS_INSTANCE_WITHOUT_REFRESH_SESSION, 'get').mockResolvedValue({
//   status: 200,
//   data: {
//     csrfTokenHeaderName: 'X-CSRF-TOKEN',
//     csrfTokenValue: 'dummy',
//   },
// });
// jest.mock('features/backend/apis/system/system', () => {
//   // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
//   const originalModule = jest.requireActual('features/backend/apis/system/system');
//   // eslint-disable-next-line @typescript-eslint/no-unsafe-return
//   return {
//     ...originalModule,
//     getAppUpdates: jest.fn(() => {
//       return Promise.resolve({
//         status: 200,
//         data: {
//           updateRequired: false,
//           supportedVersion: 'dummy',
//           message: 'dummy',
//         },
//       });
//     }),
//   };
// });
//
// jest.mock('features/backend/apis/account/account', () => {
//   // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
//   const originalModule = jest.requireActual('features/backend/apis/account/account');
//   // eslint-disable-next-line @typescript-eslint/no-unsafe-return
//   return {
//     ...originalModule,
// getAccountsMe: jest.fn(() => {
//   return Promise.resolve({
//     status: 200,
//     data: {
//       accountId: '123456789',
//       deviceTokens: [],
//     },
//   });
// }),
// getAccountsMeTerms: jest.fn(() => {
//   return Promise.resolve({
//     status: 200,
//     data: {
//       hasAgreed: true,
//       agreedVersion: '1.0.0',
//     },
//   });
// }),
//     postLogin: jest.fn(() => {
//       return Promise.resolve({
//         status: 200,
//         data: {
//           status: 'COMPLETE',
//         },
//       });
//     }),
//   };
// });
//
// jest.mock('features/backend/apis/terms/terms', () => {
//   // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
//   const originalModule = jest.requireActual('features/backend/apis/terms/terms');
//   // eslint-disable-next-line @typescript-eslint/no-unsafe-return
//   return {
//     ...originalModule,
//     getTerms: jest.fn(),
//   };
// });

const server = setupServer(...handlers);
beforeAll(async () => {
  initialDb();
  await initialData();
  server.listen();
});
afterAll(() => server.close());

beforeEach(() => {
  // 画面遷移時のアニメーションが、コンポーネントのアンマウント後にステートを更新してしまうようで、
  // テストは成功するものの、エラーログが出力されてしまう。
  // タイマーを使わないようにして、アニメーションを動かさないことで回避しているつもり。
  jest.useFakeTimers();
});

jest.mock('expo-application', () => {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
  const originalModule = jest.requireActual('expo-application');
  // eslint-disable-next-line @typescript-eslint/no-unsafe-return
  return {
    ...originalModule,
    nativeApplicationVersion: '0.1.0',
  };
});
jest.mock('expo-secure-store', () => {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
  const originalModule = jest.requireActual('expo-secure-store');
  // eslint-disable-next-line @typescript-eslint/no-unsafe-return
  return {
    ...originalModule,
    getItemAsync: jest.fn((key: string) => {
      return Promise.resolve('santoku');
    }),
  };
});

describe('App', () => {
  it('マウントされたときに正常にレンダリングされること', async () => {
    const result = render(<App />);
    await waitFor(
      () => {
        result.rerender(<App />);
        expect(screen.queryByTestId('HomePage')).not.toBeNull();
        expect(screen).toMatchSnapshot();
      },
      {timeout: 300000},
    );
  }, 300000);
});
