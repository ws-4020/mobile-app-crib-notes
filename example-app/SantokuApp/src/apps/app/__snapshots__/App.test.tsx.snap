// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App マウントされたときに正常にレンダリングされること 1`] = `
<View
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    testID="FullWindowOverlay"
  />
  <View
    testID="FullWindowOverlay"
  />
</View>
`;
